File: 02_data_formats/exercise_01/csharp/Parser.cs
using System;
using System.Text.Json;
using System.Xml.Linq;
using YamlDotNet.Serialization;
using CsvHelper;
using System.Globalization;
using Newtonsoft.Json;

namespace FileParser
{
    public class Parser
    {
        public void ParseFile(string filePath)
        {
            string fileType = Path.GetExtension(filePath);

            switch (fileType)
            {
                case ".txt":
                case ".xml":
                case ".yaml":
                case ".yml":
                case ".json":
                case ".csv":
                    ParseFileWithStreamReader(filePath, fileType);
                    break;
                default:
                    Console.WriteLine("file type: " + fileType + " not supported");
                    break;
            }
        }

        private void ParseFileWithStreamReader(string filePath, string fileType)
        {
            string fileContent;

            using (var reader = new StreamReader(filePath))
            {
                fileContent = reader.ReadToEnd();
                var person = new Person();

                switch (fileType)
                {
                    case ".txt":
                        Console.WriteLine(fileContent);
                        break;
                    case ".xml":
                        var doc = XDocument.Parse(fileContent);
                        Console.WriteLine(doc.ToString());
                        break;
                    case ".yaml":
                        var deserializer = new DeserializerBuilder().Build();
                        var result = deserializer.Deserialize(reader);
                        Console.WriteLine(result);
                        break;
                    case ".json":
                        dynamic? parsedJson = JsonConvert.DeserializeObject(fileContent);

                        person.Name = parsedJson?.name;
                        person.Age = parsedJson?.age;

                        string? s = parsedJson?.hobby;
                        string[]? hobbies = s?.Split(",");

                        var list = new List<string>();
                        if (hobbies != null)
                            foreach (string hobby in hobbies)
                            {
                                list.Add(hobby);
                            }

                        person.Hobbies = list.ToArray();

                        Console.WriteLine(person.ToString());
                        break;
                    case ".csv":
                        using (var csv = new CsvReader(reader, CultureInfo.CurrentCulture))
                        {
                            var persons = csv.GetRecords<Person>();

                            foreach (var p in persons)
                            {
                                Console.WriteLine(p);
                            }
                        }

                        break;
                }
            }
        }
    }
}

File: 02_data_formats/exercise_01/python/file_parser/parser.py
import json, yaml, csv, xml.etree.ElementTree as ET

class Parser:
    def __init__(self):
        self.supported_file_types = ["txt", "xml", "yaml", "json", "csv"]
    
    def parse_file(self, file_path):
        file_type = file_path.split(".")[-1]
        if file_type not in self.supported_file_types:
            print(f"file type not supported: {file_type}")
            return

        with open(file_path, "r") as file:
            file_content = file.read()
            if file_type == "txt":
                print(file_content)
            elif file_type == "xml":
                root = ET.fromstring(file_content)
                print(ET.tostring(root, encoding="unicode"))
            elif file_type == "yaml":
                data = yaml.safe_load(file_content)
                print(data)
            elif file_type == "json":
                data = json.loads(file_content)
                print(json.dumps(data, indent=4, default=str))
            elif file_type == "csv":
                reader = csv.reader(file_content.splitlines())
                for row in reader:
                    print(row)